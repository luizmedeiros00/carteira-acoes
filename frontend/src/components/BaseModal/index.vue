<template>
  <teleport to="body">
    <div
      v-show="active"
      class="modal fixed top-0 left-0 z-50 flex items-center justify-center w-full h-full bg-black bg-opacity-50"
      @click="clickOutSideModal"
    >
      <div
        class="modal-body overflow-hidden bg-white rounded-lg fixed animate__animated animate__fadeInDown animate__faster"
        :class="{
          'w-4/12': size === 'sm',
          'w-6/12': size === 'md',
          'w-10/12': size === 'lg',
        }"
        @click.stop
      >
        <slot name="header"></slot>
        <!-- <div v-if="title" class="flex justify-between p-4 border-b">
          <h1>{{ title }}</h1>
          <button class="text-xl text-gray-600 focus:outline-none" @click="close">&times;</button>
        </div> -->
        <slot></slot>
        <!-- <div class="p-4 gap-2">
        </div> -->
      </div>
    </div>
  </teleport>
</template>

<script setup lang="ts">
  import { ref } from 'vue'
  const emit = defineEmits(['close'])

  interface Props {
    persistent?: boolean
    size?: 'md' | 'sm' | 'lg'
  }
  const { size = 'md', persistent = false } = defineProps<Props>()
  const active = ref<boolean>(false)
  const toggle = (value: boolean) => (active.value = value)

  function clickOutSideModal() {
    if (persistent) return
    emit('close')
    toggle(false)
  }

  defineExpose({
    toggle,
  })
</script>
<style scoped></style>
