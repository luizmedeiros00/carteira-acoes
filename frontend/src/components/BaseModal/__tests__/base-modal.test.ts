import BaseModal from '../index.vue'
import { h } from 'vue'
import { shallowMount, mount, VueWrapper } from '@vue/test-utils'

describe('BaseModal', () => {
  type Instance = InstanceType<typeof BaseModal>
  let mountFunction: (options?: object) => VueWrapper<Instance>

  beforeEach(() => {
    mountFunction = (options = {}) => {
      return shallowMount(BaseModal, {
        global: {
          stubs: { teleport: true },
        },
        ...options,
      })
    }
  })

  test('should render component and match snapshot', () => {
    const wrapper = mountFunction()
    expect(wrapper.html()).toMatchSnapshot()
  })

  test('should render component and match snapshot with props size is set sm', () => {
    const wrapper = mountFunction({
      props: {
        size: 'sm',
      },
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  test('should render component and match snapshot with props size is set lg', () => {
    const wrapper = mountFunction({
      props: {
        size: 'lg',
      },
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  test('should show component when call toggle method', async () => {
    const wrapper = mountFunction({ attachTo: document.body })

    expect(wrapper.get('.modal').isVisible()).toBe(false)

    await wrapper.vm.toggle(true)
    expect(wrapper.get('.modal').isVisible()).toBe(true)
  })

  test('should close component when click outside and props persistent is set to false', async () => {
    const wrapper = mountFunction({
      attachTo: document.body,
      props: { persistent: false },
    })

    await wrapper.vm.toggle(true)
    expect(wrapper.get('.modal').isVisible()).toBe(true)

    await wrapper.find('.modal').trigger('click')
    expect(wrapper.get('.modal').isVisible()).toBe(false)
  })

  test('should not close component when click outside and props persistent is set to true', async () => {
    const wrapper = mountFunction({
      attachTo: document.body,
      props: { persistent: true },
    })

    await wrapper.vm.toggle(true)
    expect(wrapper.get('.modal').isVisible()).toBe(true)

    await wrapper.find('.modal').trigger('click')
    expect(wrapper.get('.modal').isVisible()).toBe(true)
  })

  test('should render slots', () => {
    const wrapper = mountFunction({
      attachTo: document.body,
      slots: {
        default: 'Default',
        header: '<h1>Title Modal</h1>',
      },
    })

    expect(wrapper.html()).toContain('Default')
    expect(wrapper.html()).toContain('<h1>Title Modal</h1>')
  })
})
