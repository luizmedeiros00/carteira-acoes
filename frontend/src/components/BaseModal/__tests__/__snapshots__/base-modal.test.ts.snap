// Vitest Snapshot v1

exports[`BaseModal > should render component and match snapshot 1`] = `
"<teleport-stub>
  <div class=\\"modal fixed top-0 left-0 z-50 flex items-center justify-center w-full h-full bg-black bg-opacity-50\\" style=\\"display: none;\\">
    <div class=\\"modal-body overflow-hidden bg-white rounded-lg fixed animate__animated animate__fadeInDown animate__faster\\">
      <!-- <div v-if=\\"title\\" class=\\"flex justify-between p-4 border-b\\">
          <h1>{{ title }}</h1>
          <button class=\\"text-xl text-gray-600 focus:outline-none\\" @click=\\"close\\">&times;</button>
        </div> -->
      <!-- <div class=\\"p-4 gap-2\\">
        </div> -->
    </div>
  </div>
</teleport-stub>"
`;

exports[`BaseModal > should render component and match snapshot with props size is set lg 1`] = `
"<teleport-stub>
  <div class=\\"modal fixed top-0 left-0 z-50 flex items-center justify-center w-full h-full bg-black bg-opacity-50\\" style=\\"display: none;\\">
    <div class=\\"modal-body overflow-hidden bg-white rounded-lg fixed animate__animated animate__fadeInDown animate__faster w-10/12\\">
      <!-- <div v-if=\\"title\\" class=\\"flex justify-between p-4 border-b\\">
          <h1>{{ title }}</h1>
          <button class=\\"text-xl text-gray-600 focus:outline-none\\" @click=\\"close\\">&times;</button>
        </div> -->
      <!-- <div class=\\"p-4 gap-2\\">
        </div> -->
    </div>
  </div>
</teleport-stub>"
`;

exports[`BaseModal > should render component and match snapshot with props size is set sm 1`] = `
"<teleport-stub>
  <div class=\\"modal fixed top-0 left-0 z-50 flex items-center justify-center w-full h-full bg-black bg-opacity-50\\" style=\\"display: none;\\">
    <div class=\\"modal-body overflow-hidden bg-white rounded-lg fixed animate__animated animate__fadeInDown animate__faster w-4/12\\">
      <!-- <div v-if=\\"title\\" class=\\"flex justify-between p-4 border-b\\">
          <h1>{{ title }}</h1>
          <button class=\\"text-xl text-gray-600 focus:outline-none\\" @click=\\"close\\">&times;</button>
        </div> -->
      <!-- <div class=\\"p-4 gap-2\\">
        </div> -->
    </div>
  </div>
</teleport-stub>"
`;
